### ISSUE TITLE
How to use TF2.2 with Cuda 10.2?

### ISSUE THREAD

**Issue by FelipeR2U** [tensorflow/profiler#3]
I have a multi-gpu setup to profile, so I would need to install Cuda 10.2.
But using TF 2.2.0-rc(0,1,2) I couldn't use my GPUs since TF looks for Cuda 10.1

```
2020-04-03 15:47:30.610616: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-04-03 15:47:32.728673: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-04-03 15:47:32.729831: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:00:1b.0 name: Tesla T4 computeCapability: 7.5
coreClock: 1.59GHz coreCount: 40 deviceMemorySize: 14.75GiB deviceMemoryBandwidth: 298.08GiB/s
2020-04-03 15:47:32.729940: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-04-03 15:47:32.731020: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 1 with properties: 
pciBusID: 0000:00:1c.0 name: Tesla T4 computeCapability: 7.5
coreClock: 1.59GHz coreCount: 40 deviceMemorySize: 14.75GiB deviceMemoryBandwidth: 298.08GiB/s
2020-04-03 15:47:32.731105: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-04-03 15:47:32.732204: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 2 with properties: 
pciBusID: 0000:00:1d.0 name: Tesla T4 computeCapability: 7.5
coreClock: 1.59GHz coreCount: 40 deviceMemorySize: 14.75GiB deviceMemoryBandwidth: 298.08GiB/s
2020-04-03 15:47:32.732283: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-04-03 15:47:32.733386: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 3 with properties: 
pciBusID: 0000:00:1e.0 name: Tesla T4 computeCapability: 7.5
coreClock: 1.59GHz coreCount: 40 deviceMemorySize: 14.75GiB deviceMemoryBandwidth: 298.08GiB/s
2020-04-03 15:47:32.733811: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:
2020-04-03 15:47:32.735654: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2020-04-03 15:47:32.737341: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2020-04-03 15:47:32.737683: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2020-04-03 15:47:32.739523: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2020-04-03 15:47:32.740580: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2020-04-03 15:47:32.744448: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-04-03 15:47:32.744474: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
```
How am I able to solve this and execute the profiler?

**Comment by qiuminxu** [comment]
Using CUDA 10.2 should be able to profile TF2.2 on multiple GPUs. The steps to enable this is to 1) install latest NV driver, 2) install CUDA 10.2 and the latest cuDNN library, 3) make fake symbolic links for libcudart and libcupti in /usr/local/cuda-10.2/lib64 to have the 10.1 version symbolic links (TF specifically requires 10.1 lib files for these two CUDA related libraries, not others).

To be noted, faking the symbolic links IS NOT a suggested way of using CUDA per NVIDIA's standard (the suggested way is to recompile TF with CUDA 10.2 toolchain). But that gives a simple and easy way to try whether things work without spending a lot of time figuring out the compilation steps.

If your default system come with cupti 10.1 and it is hard to upgrade. It is still likely profiling work out of box. Note that known issues with cupti 10.1 only happens when profiling multiple GPUs. If you only profile a single GPU, that would be fine.

If the profiling is not successful, try set the following env variable:
export TF_GPU_CUPTI_USE_ACTIVITY_API=false
This will also allow you to profile multiple GPUs, but potentially incur higher overhead to GPU execution time while tracing. 

Let me know if you can profile successfully!

**Comment by FelipeR2U** [comment]
Thank you, I'm gonna give it a try

**Comment by FelipeR2U** [comment]
> Using CUDA 10.2 should be able to profile TF2.2 on multiple GPUs. The steps to enable this is to 1) install latest NV driver, 2) install CUDA 10.2 and the latest cuDNN library, 3) make fake symbolic links for libcudart and libcupti in /usr/local/cuda-10.2/lib64 to have the 10.1 version symbolic links (TF specifically requires 10.1 lib files for these two CUDA related libraries, not others).
> 
> To be noted, faking the symbolic links IS NOT a suggested way of using CUDA per NVIDIA's standard (the suggested way is to recompile TF with CUDA 10.2 toolchain). But that gives a simple and easy way to try whether things work without spending a lot of time figuring out the compilation steps.
> 
> If your default system come with cupti 10.1 and it is hard to upgrade. It is still likely profiling work out of box. Note that known issues with cupti 10.1 only happens when profiling multiple GPUs. If you only profile a single GPU, that would be fine.
> 
> If the profiling is not successful, try set the following env variable:
> export TF_GPU_CUPTI_USE_ACTIVITY_API=false
> This will also allow you to profile multiple GPUs, but potentially incur higher overhead to GPU execution time while tracing.
> 
> Let me know if you can profile successfully!

Making a fake symbolic link worked to tensorflow 2.2 finding the libraries, but the training process never starts after a `OP_REQUIRES failed at cast_op.cc:123 : Unimplemented: Cast string to float is not supported` warning that does not happen in TF 2.1.

**Comment by qiuminxu** [comment]
Is it possible to share your model code so that I can replicate the issue? Thanks.

**Comment by qiuminxu** [comment]
Hi @FelipeR2U 

I added a guide on how to use TF 2.2 with Cuda 10.2 here. Would you take a look and see if it works?
https://github.com/tensorflow/profiler/blob/master/docs/profile_multi_gpu.md

**Comment by FelipeR2U** [comment]
I traced my problem to what seems to be the loss function. I'm using the one from TensorflowAddons TripletSemiHardLoss.
I'm going to try your guide and tell the results.

> Hi @FelipeR2U
> 
> I added a guide on how to use TF 2.2 with Cuda 10.2 here. Would you take a look and see if it works?
> https://github.com/tensorflow/profiler/blob/master/docs/profile_multi_gpu.md
